lib/client/alpn.go

```diff
@@ -0,0 +1,233 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/keys"
+	"github.com/gravitational/teleport/lib/srv/alpnproxy"
+	alpn "github.com/gravitational/teleport/lib/srv/alpnproxy/common"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// ALPNAuthTunnel contains the required methods to create a local ALPN proxy.
+type ALPNAuthTunnel interface {
+	// GetClusterCACert returns the PEM-encoded TLS certs for the local cluster.
+	// If the cluster has multiple TLS certs, they will all be concatenated.
+	GetClusterCACert(ctx context.Context) (*proto.GetClusterCACertResponse, error)
+
+	// GetCurrentUser returns current user as seen by the server.
+	// Useful especially in the context of remote clusters which perform role and trait mapping.
+	GetCurrentUser(ctx context.Context) (types.User, error)
+
+	// GenerateUserCerts takes the public key in the OpenSSH `authorized_keys` plain
+	// text format, signs it using User Certificate Authority signing key and
+	// returns the resulting certificates.
+	GenerateUserCerts(ctx context.Context, req proto.UserCertsRequest) (*proto.Certs, error)
+}
+
+// RunALPNAuthTunnelRequest contains the required fields used to create an authed ALPN Proxy
+type RunALPNAuthTunnelRequest struct {
+	// Client is the client that's used to interact with the cluster and obtain Certificates.
+	Client ALPNAuthTunnel
+
+	// Listener to be used to accept connections that will go trough the
+	Listener net.Listener
+
+	// Insecure turns off verification for x509 upstream ALPN proxy service certificate.
+	Insecure bool
+
+	// Protocol name
+	// Examples for databases: "postgres", "mysql"
+	// This protocol must map to a Teleport ALPN protocol [lib/srv/alpnproxy/common.alpnToALPNProtocol]
+	Protocol string
+
+	// WebProxyAddr is the proxy addr to
+	WebProxyAddr string
+
+	// ConnectionDiagnosticID contains the ID to be used to store Connection Diagnostic checks.
+	// Can be empty.
+	ConnectionDiagnosticID string
+
+	// ProxyMiddleware is a middleware that ensures that the local proxy has valid TLS certs.
+	ProxyMiddleware alpnproxy.LocalProxyMiddleware
+
+	// RouteToDatabase contains the destination server that must receive the connection.
+	// Specific for database proxying.
+	RouteToDatabase proto.RouteToDatabase
+}
+
+// RunALPNAuthTunnel runs a local authenticated ALPN proxy to another service.
+// At least one Route (which defines the service) must be defined
+func RunALPNAuthTunnel(ctx context.Context, req RunALPNAuthTunnelRequest) error {
+	alpnProtocol, err := alpn.ToALPNProtocol(req.Protocol)
```

Comment:
nit: alternatively just get the database protocol from `RouteToDatabase`

Reply:
This was meant to be generic and not depend on being a proxy to a Database
We can add a RouteToApp later